File: polymarket_predict_bot/batch_order.py
"""
批量快速挂单脚本 - Predict.fun 有星星的市场
功能:
  - 获取所有有星星(有奖励积分)的市场
  - 体育比赛: 未开赛且距开赛>30分钟 → 挂单; 已开赛或<30分钟 → 跳过
  - 每个市场以买1价格挂10份额
  - 总预算控制30U
  - 买1买2数量级匹配检查

使用:
  python3 batch_order.py
"""

import requests
import logging
import time
import re
from datetime import datetime, timedelta, timezone

from config import (
    PREDICT_API_URL,
    PREDICT_API_KEY,
    ORDER_SIZE,
    TOTAL_BUDGET,
    ONLY_WITH_REWARDS,
    SKIP_LIVE_EVENTS,
)

# 尝试导入 JWT Token (挂单用)
try:
    from config import PREDICT_JWT_TOKEN
except ImportError:
    PREDICT_JWT_TOKEN = ""

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S",
)
logger = logging.getLogger(__name__)


class PredictBatchTrader:
    """Predict.fun 批量挂单"""

    def __init__(self):
        self.base_url = PREDICT_API_URL
        # 读取用 session (x-api-key)
        self.session = requests.Session()
        self.session.headers.update({
            "Content-Type": "application/json",
            "x-api-key": PREDICT_API_KEY,
        })
        # 挂单用 session (Bearer JWT)
        jwt = PREDICT_JWT_TOKEN if PREDICT_JWT_TOKEN and PREDICT_JWT_TOKEN != "YOUR_JWT_TOKEN_HERE" else PREDICT_API_KEY
        self.order_session = requests.Session()
        self.order_session.headers.update({
            "Authorization": f"Bearer {jwt}",
            "Content-Type": "application/json",
            "x-api-key": PREDICT_API_KEY,
        })
        self.total_spent = 0.0

    def get_markets(self):
        """
        获取市场列表
        API: GET /v1/markets
        """
        url = f"{self.base_url}/v1/markets"
        params = {
            "status": "OPEN",
            "first": 100,
        }
        if ONLY_WITH_REWARDS:
            params["hasActiveRewards"] = "true"

        try:
            resp = self.session.get(url, params=params, timeout=10)
            resp.raise_for_status()
            data = resp.json()

            markets = data.get("data", [])
            if not markets:
                markets = data if isinstance(data, list) else []

            logger.info(f"获取到 {len(markets)} 个市场")
            return markets

        except requests.exceptions.RequestException as e:
            logger.error(f"获取市场列表失败: {e}")
            if hasattr(e, "response") and e.response is not None:
                logger.error(f"响应: {e.response.text}")
            return []

    def is_live_event(self, market):
        """
        判断市场是否已经开始/进行中
        跳过正在比赛中的事件
        """
        # 检查 tradingStatus
        trading_status = market.get("tradingStatus", "")
        if isinstance(trading_status, dict):
            trading_status = trading_status.get("status", "")

        # 常见的进行中状态
        live_statuses = ["LIVE", "IN_PROGRESS", "STARTED", "HALTED"]
        if str(trading_status).upper() in live_statuses:
            return True

        # 检查标题中是否有 "LIVE" 标记
        title = market.get("title", "") or market.get("question", "")
        if "[LIVE]" in title or "(LIVE)" in title:
            return True

        return False

    def parse_event_start_time(self, market):
        """
        从市场的 description 或 categorySlug 解析开赛时间
        返回 datetime (UTC) 或 None
        
        支持格式:
          - description 中: "scheduled for May 5 at 7:45PM ET"
          - categorySlug 中: "mlb-col-pit-2026-05-12" (只有日期，无具体时间)
        """
        desc = market.get("description", "") or ""
        cat = market.get("categorySlug", "") or ""
        
        # 方法1: 从 description 解析 "scheduled for May 12 at 7:05PM ET"
        # 或 "initially scheduled for May 14 at 11:00AM ET"
        pattern = r'scheduled for (\w+ \d+)(?: at (\d+:\d+(?:AM|PM)) ET)?'
        match = re.search(pattern, desc, re.IGNORECASE)
        if match:
            date_str = match.group(1)  # "May 12"
            time_str = match.group(2)  # "7:05PM" 或 None
            
            # 从 categorySlug 中取年份
            year_match = re.search(r'(\d{4})-(\d{2})-(\d{2})$', cat)
            if year_match:
                year = int(year_match.group(1))
            else:
                year = datetime.now().year
            
            try:
                if time_str:
                    dt_str = f"{date_str} {year} {time_str}"
                    dt = datetime.strptime(dt_str, "%B %d %Y %I:%M%p")
                else:
                    dt_str = f"{date_str} {year}"
                    dt = datetime.strptime(dt_str, "%B %d %Y")
                
                # ET = UTC-4 (EDT) 或 UTC-5 (EST)，5月份是 EDT (UTC-4)
                et_offset = timedelta(hours=-4)
                dt_utc = dt - et_offset  # 转为 UTC
                return dt_utc
            except ValueError:
                pass
        
        # 方法2: 从 categorySlug 解析日期 (无具体时间，默认当天 23:59 ET)
        date_match = re.search(r'(\d{4})-(\d{2})-(\d{2})$', cat)
        if date_match:
            try:
                year = int(date_match.group(1))
                month = int(date_match.group(2))
                day = int(date_match.group(3))
                # 没有具体时间，不做开赛前30分钟撤单（因为不知道具体时间）
                return None
            except ValueError:
                pass
        
        return None

    def should_skip_sports_market(self, market):
        """
        体育比赛市场是否应该跳过:
        - 已开赛(LIVE) → 跳过
        - 距开赛 ≤ 30分钟 → 跳过
        - 距开赛 > 30分钟 → 不跳过(可以挂)
        - 无法解析开赛时间 → 不跳过(可以挂)
        
        返回: (should_skip: bool, reason: str)
        """
        # 先检查是否已经 LIVE
        if self.is_live_event(market):
            return True, "已开赛"
        
        # 解析开赛时间
        start_time = self.parse_event_start_time(market)
        if start_time is None:
            # 无法解析开赛时间，允许挂单
            return False, ""
        
        now_utc = datetime.utcnow()
        time_to_start = start_time - now_utc
        minutes_to_start = time_to_start.total_seconds() / 60
        
        if minutes_to_start <= 0:
            return True, "已开赛"
        elif minutes_to_start <= 30:
            return True, f"距开赛仅{minutes_to_start:.0f}分钟(<30min)"
        else:
            return False, ""

    def has_rewards(self, market):
        """判断市场是否有星星(奖励积分)"""
        rewards = market.get("rewards")
        if rewards is None:
            return False
        if isinstance(rewards, dict):
            current = rewards.get("current")
            schedule = rewards.get("schedule", [])
            return current is not None or len(schedule) > 0
        return True

    def _parse_bid(self, bid):
        """解析单个 bid 的价格和数量"""
        if isinstance(bid, dict):
            price = float(bid.get("price", 0))
            size = float(bid.get("size", 0))
        elif isinstance(bid, list):
            price = float(bid[0])
            size = float(bid[1]) if len(bid) > 1 else 0
        else:
            price = float(bid)
            size = 0
        return price, size

    def check_magnitude_match(self, bid1_size, bid2_size):
        """
        买1买2数量级匹配检查:
        - 买2≥10000 则买1也要≥10000
        - 买2≥1000 则买1也要≥1000
        - 买2≥100 则买1也要≥100
        - 买2≥10 则买1也要≥10
        - 以此类推
        不匹配则返回 False（不挂单）
        """
        # 从大到小检查数量级
        thresholds = [10000, 1000, 100, 10]
        for threshold in thresholds:
            if bid2_size >= threshold:
                if bid1_size < threshold:
                    logger.warning(
                        f"  ⚠️ 买1买2数量级不匹配! "
                        f"买2={bid2_size:.0f}≥{threshold}, "
                        f"但买1={bid1_size:.0f}<{threshold}, 跳过"
                    )
                    return False
                return True
        # 买2 < 10，不做限制
        return True

    def get_market_orderbook(self, market_id):
        """
        获取指定市场的盘口
        API: GET /v1/markets/{id}/orderbook
        """
        url = f"{self.base_url}/v1/markets/{market_id}/orderbook"

        try:
            resp = self.session.get(url, timeout=10)
            resp.raise_for_status()
            data = resp.json()

            orderbook = data.get("data", data)
            bids = orderbook.get("bids", [])

            if bids:
                bid1_price, bid1_size = self._parse_bid(bids[0])

                # 解析买2
                bid2_size = 0
                if len(bids) > 1:
                    _, bid2_size = self._parse_bid(bids[1])

                return {
                    "bid1_price": bid1_price,
                    "bid1_size": bid1_size,
                    "bid2_size": bid2_size,
                    "bids": bids,
                }
            else:
                return None

        except requests.exceptions.RequestException as e:
            logger.error(f"获取盘口失败 (market_id={market_id}): {e}")
            return None
        except (IndexError, ValueError, TypeError) as e:
            logger.error(f"解析盘口失败 (market_id={market_id}): {e}")
            return None

    def create_order(self, market_id, side, price, size):
        """
        创建限价单
        API: POST /v1/orders
        使用 JWT Token 认证
        """
        url = f"{self.base_url}/v1/orders"

        payload = {
            "marketId": market_id,
            "side": side,
            "price": str(round(price, 4)),
            "size": str(round(size, 2)),
            "strategy": "LIMIT",
        }

        try:
            resp = self.order_session.post(url, json=payload, timeout=10)
            resp.raise_for_status()
            order = resp.json()
            logger.info(
                f"  ✅ 挂单成功: {side} @ {price:.4f} x {size}"
            )
            return order
        except requests.exceptions.RequestException as e:
            logger.error(f"  ❌ 挂单失败: {e}")
            if hasattr(e, "response") and e.response is not None:
                logger.error(f"  响应: {e.response.text}")
            return None

    def batch_place_orders(self):
        """
        批量挂单:
        - 只挂有星星的市场
        - 跳过已开始的比赛
        - 每个市场买1挂10份额
        - 总预算30U
        """
        logger.info("=" * 50)
        logger.info("Predict.fun 批量挂单 (30U 测试)")
        logger.info(f"每笔份额: {ORDER_SIZE}")
        logger.info(f"总预算: {TOTAL_BUDGET} USDB")
        logger.info(f"只挂有星星的: {ONLY_WITH_REWARDS}")
        logger.info(f"跳过进行中: {SKIP_LIVE_EVENTS}")
        logger.info("=" * 50)

        # 1. 获取市场
        markets = self.get_markets()
        if not markets:
            logger.error("没有找到可交易的市场!")
            return

        success_count = 0
        fail_count = 0
        skip_count = 0

        # 2. 逐个市场处理
        for i, market in enumerate(markets, 1):
            # 检查预算
            if self.total_spent >= TOTAL_BUDGET:
                logger.info(f"\n💰 已达到预算上限 {TOTAL_BUDGET} USDB，停止挂单")
                break

            market_id = market.get("id") or market.get("marketId")
            market_name = (
                market.get("title")
                or market.get("question")
                or market.get("name")
                or f"Market #{market_id}"
            )

            # 过滤: 只挂有星星的
            if ONLY_WITH_REWARDS and not self.has_rewards(market):
                continue

            # 过滤: 跳过已开始的比赛 / 开赛前30分钟内的比赛
            market_variant = market.get("marketVariant", "")
            is_sports = market_variant in ["SPORTS_MATCH", "SPORTS_TEAM_MATCH"]
            
            if is_sports:
                should_skip, skip_reason = self.should_skip_sports_market(market)
                if should_skip:
                    logger.info(f"[{i}] ⏭️ 跳过({skip_reason}): {market_name}")
                    skip_count += 1
                    continue
            elif SKIP_LIVE_EVENTS and self.is_live_event(market):
                logger.info(f"[{i}] ⏭️ 跳过(进行中): {market_name}")
                skip_count += 1
                continue

            logger.info(f"\n[{i}] {market_name}")

            # 获取盘口
            book = self.get_market_orderbook(market_id)
            if book is None or book["bid1_price"] <= 0:
                logger.warning(f"  跳过: 没有买盘")
                skip_count += 1
                continue

            bid1_price = book["bid1_price"]
            bid1_size = book["bid1_size"]
            bid2_size = book.get("bid2_size", 0)
            logger.info(f"  买1: {bid1_price:.4f} (量={bid1_size:.0f}), 买2量={bid2_size:.0f}")

            # 买1买2数量级匹配检查
            if not self.check_magnitude_match(bid1_size, bid2_size):
                skip_count += 1
                continue

            # 计算这笔单花多少钱
            cost = bid1_price * ORDER_SIZE
            if self.total_spent + cost > TOTAL_BUDGET:
                logger.info(f"  跳过: 预算不足 (剩余{TOTAL_BUDGET - self.total_spent:.2f})")
                break

            # 挂单
            result = self.create_order(
                market_id=market_id,
                side="BUY",
                price=bid1_price,
                size=ORDER_SIZE,
            )

            if result:
                success_count += 1
                self.total_spent += cost
                logger.info(f"  花费: {cost:.2f} USDB (累计: {self.total_spent:.2f}/{TOTAL_BUDGET})")
            else:
                fail_count += 1

            # 控制速度
            time.sleep(0.5)

        # 3. 汇总
        logger.info("\n" + "=" * 50)
        logger.info(f"批量挂单完成!")
        logger.info(f"✅ 成功: {success_count} 笔")
        logger.info(f"❌ 失败: {fail_count} 笔")
        logger.info(f"⏭️ 跳过: {skip_count} 笔")
        logger.info(f"💰 总花费: {self.total_spent:.2f} USDB")
        logger.info("=" * 50)


def main():
    if PREDICT_API_KEY == "YOUR_PREDICT_FUN_JWT_TOKEN_HERE":
        print("\n⚠️  请先配置 config.py 中的 PREDICT_API_KEY!")
        return

    trader = PredictBatchTrader()
    trader.batch_place_orders()


if __name__ == "__main__":
    main()

File: polymarket_predict_bot/batch_order.js
/**
 * 批量快速挂单脚本 - Predict.fun 有星星的市场
 * 
 * 功能:
 *   - 获取所有有星星(有奖励积分)的市场
 *   - 跳过已经开始/进行中的比赛
 *   - 每个市场以买1价格挂10份额
 *   - 总预算控制30U
 *   - 成交后 Telegram 通知
 * 
 * 使用:
 *   1. npm install
 *   2. 编辑下方配置
 *   3. node batch_order.js
 */

import { Wallet } from "ethers";
import { OrderBuilder, ChainId, Side } from "@predictdotfun/sdk";

// ============ 配置 ============
const CONFIG = {
  // 你的钱包私钥 (从 MetaMask 导出, 不要分享给任何人!)
  PRIVATE_KEY: "YOUR_PRIVATE_KEY_HERE",

  // Predict.fun API Key (UUID格式)
  API_KEY: "5f623dc1-147a-4767-8795-cf02f1f25149",

  // Predict.fun JWT Token (从浏览器获取)
  JWT_TOKEN: "YOUR_JWT_TOKEN_HERE",

  // API URL
  API_URL: "https://api.predict.fun",

  // 交易参数
  TOTAL_BUDGET: 30.0,     // 总预算 30 USDB
  ORDER_SIZE: 10,         // 每个市场挂 10 份额
  SIDE: Side.BUY,         // 买入方向 (0=BUY, 1=SELL)

  // 过滤
  ONLY_WITH_REWARDS: true,  // 只挂有星星的
  SKIP_LIVE_EVENTS: true,   // 跳过进行中的比赛

  // Telegram 通知
  TELEGRAM_BOT_TOKEN: "8739215233:AAHwG7G60sgOYze9Jo0u-KddtP0UBxDjnKg",
  TELEGRAM_CHAT_ID: "5707621530",
};

// ============ 工具函数 ============

async function sendTelegram(message) {
  if (!CONFIG.TELEGRAM_BOT_TOKEN || CONFIG.TELEGRAM_BOT_TOKEN === "YOUR_TELEGRAM_BOT_TOKEN_HERE") return;
  try {
    const url = `https://api.telegram.org/bot${CONFIG.TELEGRAM_BOT_TOKEN}/sendMessage`;
    const resp = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ chat_id: CONFIG.TELEGRAM_CHAT_ID, text: message, parse_mode: "HTML" }),
    });
    if (resp.ok) console.log("[通知] Telegram 发送成功");
  } catch (e) {
    console.error("[通知] Telegram 发送失败:", e.message);
  }
}

async function fetchAPI(path, options = {}) {
  const url = `${CONFIG.API_URL}${path}`;
  const headers = {
    "Content-Type": "application/json",
    "x-api-key": CONFIG.API_KEY,
    "Authorization": `Bearer ${CONFIG.JWT_TOKEN}`,
    ...options.headers,
  };
  const resp = await fetch(url, { ...options, headers });
  if (!resp.ok) {
    const text = await resp.text();
    throw new Error(`API ${resp.status}: ${text}`);
  }
  return resp.json();
}

// ============ 获取市场 ============

async function getMarkets() {
  const params = new URLSearchParams({
    status: "OPEN",
    first: "100",
    hasActiveRewards: "true",
  });
  const data = await fetchAPI(`/v1/markets?${params}`);
  return data.data || [];
}

// ============ 获取盘口 ============

async function getOrderbook(marketId) {
  const data = await fetchAPI(`/v1/markets/${marketId}/orderbook`);
  const orderbook = data.data || data;
  const bids = orderbook.bids || [];
  if (bids.length === 0) return null;

  const firstBid = bids[0];
  let price, size;
  if (Array.isArray(firstBid)) {
    price = parseFloat(firstBid[0]);
    size = parseFloat(firstBid[1] || 0);
  } else if (typeof firstBid === "object") {
    price = parseFloat(firstBid.price || 0);
    size = parseFloat(firstBid.size || 0);
  } else {
    price = parseFloat(firstBid);
    size = 0;
  }
  return { price, size };
}

// ============ 判断是否进行中 ============

function isLiveEvent(market) {
  const status = typeof market.tradingStatus === "object"
    ? market.tradingStatus?.status || ""
    : market.tradingStatus || "";
  const liveStatuses = ["LIVE", "IN_PROGRESS", "STARTED", "HALTED"];
  if (liveStatuses.includes(status.toUpperCase())) return true;

  const title = market.title || market.question || "";
  if (title.includes("[LIVE]") || title.includes("(LIVE)")) return true;

  return false;
}

// ============ 判断是否有星星 ============

function hasRewards(market) {
  const rewards = market.rewards;
  if (!rewards) return false;
  if (rewards.current || (rewards.schedule && rewards.schedule.length > 0)) return true;
  return false;
}

// ============ 获取 tokenId ============

function getTokenId(market) {
  // 从 market outcomes 获取 Yes 的 tokenId
  const outcomes = market.outcomes || [];
  for (const outcome of outcomes) {
    if (typeof outcome === "object") {
      // 可能的字段名: tokenId, id, onChainId, token_id
      const tid = outcome.tokenId || outcome.onChainId || outcome.token_id || outcome.id;
      if (tid) return String(tid);
    } else if (typeof outcome === "string") {
      return outcome;
    }
  }
  // 从 conditionId 获取
  if (market.conditionId) return market.conditionId;
  // 从 oracleQuestionId 获取
  if (market.oracleQuestionId) return market.oracleQuestionId;
  return null;
}

// 调试: 打印第一个市场的完整数据结构
function debugMarket(market) {
  console.log("\n[DEBUG] 市场数据结构:");
  console.log("  id:", market.id);
  console.log("  title:", market.title);
  console.log("  conditionId:", market.conditionId);
  console.log("  oracleQuestionId:", market.oracleQuestionId);
  console.log("  outcomes:", JSON.stringify(market.outcomes, null, 2));
  console.log("  isNegRisk:", market.isNegRisk);
  console.log("  isYieldBearing:", market.isYieldBearing);
  console.log("  feeRateBps:", market.feeRateBps);
}

// ============ 主函数 ============

async function main() {
  console.log("=".repeat(50));
  console.log("Predict.fun 批量挂单 (Node.js + 官方SDK)");
  console.log(`预算: ${CONFIG.TOTAL_BUDGET} USDB`);
  console.log(`每笔: ${CONFIG.ORDER_SIZE} 份额`);
  console.log(`只挂有星星: ${CONFIG.ONLY_WITH_REWARDS}`);
  console.log(`跳过进行中: ${CONFIG.SKIP_LIVE_EVENTS}`);
  console.log("=".repeat(50));

  // 检查私钥
  if (CONFIG.PRIVATE_KEY === "YOUR_PRIVATE_KEY_HERE") {
    console.error("\n⚠️  请在 batch_order.js 顶部 CONFIG 中填入你的钱包私钥!");
    console.error("从 MetaMask → 账户详情 → 显示私钥 获取");
    return;
  }

  // 初始化钱包和 OrderBuilder
  console.log("\n初始化钱包和 SDK...");
  const signer = new Wallet(CONFIG.PRIVATE_KEY);
  console.log(`钱包地址: ${signer.address}`);

  const orderBuilder = await OrderBuilder.make(ChainId.BnbMainnet, signer);
  console.log("SDK 初始化成功!");

  // 设置 approvals (首次需要, 之后可以跳过)
  console.log("\n检查/设置 approvals...");
  try {
    const approvalResult = await orderBuilder.setApprovals();
    if (approvalResult.success) {
      console.log("Approvals 已设置!");
    } else {
      console.warn("Approvals 设置失败, 可能已经设置过了, 继续...");
    }
  } catch (e) {
    console.warn("Approvals 检查异常 (可能已设置):", e.message);
  }

  // 获取市场列表
  console.log("\n获取市场列表...");
  const markets = await getMarkets();
  console.log(`获取到 ${markets.length} 个市场`);

  let totalSpent = 0;
  let successCount = 0;
  let failCount = 0;
  let skipCount = 0;

  // 逐个市场处理
  // 先调试打印第一个市场数据
  if (markets.length > 0) {
    debugMarket(markets[0]);
  }

  for (let i = 0; i < markets.length; i++) {
    const market = markets[i];

    // 预算检查
    if (totalSpent >= CONFIG.TOTAL_BUDGET) {
      console.log(`\n💰 已达预算上限 ${CONFIG.TOTAL_BUDGET} USDB, 停止`);
      break;
    }

    const marketId = market.id || market.marketId;
    const marketName = market.title || market.question || `Market #${marketId}`;

    // 过滤: 只挂有星星的
    if (CONFIG.ONLY_WITH_REWARDS && !hasRewards(market)) continue;

    // 过滤: 跳过进行中
    if (CONFIG.SKIP_LIVE_EVENTS && isLiveEvent(market)) {
      console.log(`[${i + 1}] ⏭️ 跳过(进行中): ${marketName}`);
      skipCount++;
      continue;
    }

    console.log(`\n[${i + 1}] ${marketName}`);

    // 获取盘口
    let book;
    try {
      book = await getOrderbook(marketId);
    } catch (e) {
      console.warn(`  获取盘口失败: ${e.message}`);
      skipCount++;
      continue;
    }

    if (!book || book.price <= 0) {
      console.log("  跳过: 没有买盘");
      skipCount++;
      continue;
    }

    console.log(`  买1: ${book.price.toFixed(4)} (量=${book.size.toFixed(2)})`);

    // 计算花费
    const cost = book.price * CONFIG.ORDER_SIZE;
    if (totalSpent + cost > CONFIG.TOTAL_BUDGET) {
      console.log(`  跳过: 预算不足 (剩余${(CONFIG.TOTAL_BUDGET - totalSpent).toFixed(2)})`);
      break;
    }

    // 获取 tokenId
    const tokenId = getTokenId(market);
    if (!tokenId) {
      console.log("  跳过: 没有 tokenId");
      skipCount++;
      continue;
    }

    // 构建并签名订单
    try {
      const priceWei = BigInt(Math.floor(book.price * 1e18));
      const quantityWei = BigInt(CONFIG.ORDER_SIZE) * BigInt(1e18);

      const { makerAmount, takerAmount, pricePerShare } = orderBuilder.getLimitOrderAmounts({
        side: CONFIG.SIDE,
        pricePerShareWei: priceWei,
        quantityWei: quantityWei,
      });

      const order = orderBuilder.buildOrder("LIMIT", {
        maker: signer.address,
        signer: signer.address,
        side: CONFIG.SIDE,
        tokenId: tokenId,
        makerAmount,
        takerAmount,
        nonce: 0n,
        feeRateBps: market.feeRateBps || 0,
      });

      const isNegRisk = market.isNegRisk || false;
      const isYieldBearing = market.isYieldBearing || false;

      const typedData = orderBuilder.buildTypedData(order, { isNegRisk, isYieldBearing });
      const signedOrder = await orderBuilder.signTypedDataOrder(typedData);
      const hash = orderBuilder.buildTypedDataHash(typedData);

      // 提交到 API (需要将 BigInt 转为 string)
      const serializableOrder = {};
      for (const [key, value] of Object.entries(signedOrder)) {
        serializableOrder[key] = typeof value === "bigint" ? value.toString() : value;
      }
      serializableOrder.hash = hash;

      const createOrderBody = {
        data: {
          order: serializableOrder,
          pricePerShare: typeof pricePerShare === "bigint" ? pricePerShare.toString() : pricePerShare,
          strategy: "LIMIT",
        },
      };

      const result = await fetchAPI("/v1/orders", {
        method: "POST",
        body: JSON.stringify(createOrderBody),
      });

      console.log(`  ✅ 挂单成功! orderId=${result.data?.orderId || "ok"}`);
      successCount++;
      totalSpent += cost;
      console.log(`  花费: ${cost.toFixed(2)} USDB (累计: ${totalSpent.toFixed(2)}/${CONFIG.TOTAL_BUDGET})`);

      // Telegram 通知
      await sendTelegram(
        `✅ <b>挂单成功</b>\n市场: ${marketName}\n价格: ${book.price.toFixed(4)}\n数量: ${CONFIG.ORDER_SIZE}\n花费: ${cost.toFixed(2)} USDB`
      );

    } catch (e) {
      console.error(`  ❌ 挂单失败: ${e.message}`);
      failCount++;
    }

    // 控制速度
    await new Promise(r => setTimeout(r, 500));
  }

  // 汇总
  console.log("\n" + "=".repeat(50));
  console.log("批量挂单完成!");
  console.log(`✅ 成功: ${successCount} 笔`);
  console.log(`❌ 失败: ${failCount} 笔`);
  console.log(`⏭️ 跳过: ${skipCount} 笔`);
  console.log(`💰 总花费: ${totalSpent.toFixed(2)} USDB`);
  console.log("=".repeat(50));

  // 总结通知
  if (successCount > 0) {
    await sendTelegram(
      `📊 <b>批量挂单完成</b>\n✅ 成功: ${successCount}\n❌ 失败: ${failCount}\n💰 花费: ${totalSpent.toFixed(2)} USDB`
    );
  }
}

main().catch(e => {
  console.error("程序异常:", e);
  process.exit(1);
});
